src/modules/schema.js

```diff
@@ -0,0 +1,240 @@
+      validateBlobDIDHexIdentifier(did);
+      return did;
+    } catch (e) {
+      // Cannot parse as hex
+      throw new Error(`Invalid hexadecimal DID ${did}. ${e}`);
+    }
+  }
+}
+
+/**
+ * Generates a random schema ID
+ * @returns {string}
+ */
+export function createNewSchemaID() {
+  const hexId = randomAsHex(BLOB_ID_MAX_BYTE_SIZE);
+  const ss58Id = encodeAddress(hexId);
+  return `${BlobQualifier}${ss58Id}`;
+}
+
+export default class Schema {
+  /**
+   * Creates a new `Schema` object
+   * @constructor
+   * @param {string} [id] - optional schema ID, if not given, generate a random id
+   */
+  constructor(id) {
+    this.id = id || createNewSchemaID();
+    this.name = '';
+    this.version = '1.0.0';
+  }
+
+  /**
+   * Add the JSON schema to this object after checking that `json` is a valid JSON schema. Check if JSON is valid.
+   * @param {object} json - the schema JSON
+   */
+  async setJSONSchema(json) {
+    await Schema.validateSchema(json);
+    this.schema = json;
+    return this;
+  }
+
+  setName(name) {
+    this.name = name;
+    return this;
+  }
+
+  /**
+   * Update the object with `author` key. Repeatedly calling it will keep resetting the author
+   * did can be a DID hex identifier or full DID
+   * @param {string} did - the author DID
+   */
+  setAuthor(did) {
+    // TODO: `did` should be validated, do a best effort. Check either 32 byte (use constant) hex or a valid Dock DID or starts with 'did'
+    this.author = did;
+    return this;
+  }
+
+  /**
+   * Update the object with `signature` key. This method is used when
+   * signing key/capability is not present and the signature is received from outside.
+   * Repeatedly calling it will keep resetting the `signature` key.
+   * The signature must be one of the supported objects
+   * @param {object} signature - The schema's signature
+   */
+  setSignature(signature) {
+    if (signature instanceof Signature) {
+      this.signature = signature;
+    } else {
+      throw new Error('Provided signature object is not of instance Signature');
+    }
+
+    return this;
+  }
+
+  /**
+   * Serializes the object using `getSerializedBlob` and then signs it using the given
+   * polkadot-js pair. The object will be updated with key `signature`. Repeatedly calling it will
+   * keep resetting the `signature` key
+   * @param {any} msg - The message to sign
+   * @param {object} pair - Key pair to sign with
+   */
+  sign(msg, pair) {
+    this.signature = getSignatureFromKeyringPair(pair, msg);
+    return this;
+  }
+
+  /**
+   * Serializes to JSON for sending to the node, as `Blob`. A full DID is converted to the
+   * hex identifier and signature object is converted to the enum
+   * @returns {object}
+   */
+  toJSON() {
+    const {
+      signature,
+      ...rest
+    } = this;
+    return rest;
+  }
+
+  /**
+   * Serializes the schema to a blob object to send to the node
+   * @returns {object}
+   */
+  toBlob(id, did) {
```

Comment:
This function here seems weird. If you want to have a function that creates a blob from a JSON schema, have a detached one or have static in Blob like `fromJSONSchema`
Whats the point of passing id? This object will always have an id? Why allow it to be changed.

Reply:
`Whats the point of passing id? This object will always have an id? Why allow it to be changed.` ID of blob is unknown from the Schema object unless its generated externally

Reply:
A `Schema` object will always have an id, if you don't pass in the constructor, it will be auto generated

Reply:
yeah not sure why i wrote it this way for the example, have resolved